modin/core/dataframe/pandas/dataframe/dataframe.py

```diff
@@ -2535,31 +2565,66 @@ def concat(
             left_parts = self._partitions
             right_parts = [o._partitions for o in others]
-            new_lengths = self._row_lengths
-            new_widths = self._column_widths + [
-                length for o in others for length in o._column_widths
-            ]
+            new_lengths = self._row_lengths_cache
         else:
-            left_parts, right_parts, joined_index = self._copartition(
+            (
+                left_parts,
+                right_parts,
+                joined_index,
+                partition_sizes_along_axis,
+            ) = self._copartition(
                 axis.value ^ 1, others, how, sort, force_repartition=False
             )
-            new_lengths = None
-            new_widths = None
+            if axis == Axis.COL_WISE:
+                new_lengths = partition_sizes_along_axis
+            else:
+                new_widths = partition_sizes_along_axis
+
         new_partitions = self._partition_mgr_cls.concat(
             axis.value, left_parts, right_parts
         )
+
         if axis == Axis.ROW_WISE:
             new_index = self.index.append([other.index for other in others])
             new_columns = joined_index
             # TODO: Can optimize by combining if all dtypes are materialized
             new_dtypes = None
+            # If we have already cached the length of each row in at least one
+            # of the row's partitions, we can build new_lengths for the new
+            # frame.
+            new_lengths = []
+            for row in new_partitions:
+                found_row_length = False
+                for partition in row:
+                    if partition._length_cache is not None:
+                        new_lengths.append(partition.length())
+                        found_row_length = True
+                        break
+                if not found_row_length:
+                    new_lengths = None
+                    break
         else:
             new_columns = self.columns.append([other.columns for other in others])
             new_index = joined_index
             if self._dtypes is not None and all(o._dtypes is not None for o in others):
                 new_dtypes = self.dtypes.append([o.dtypes for o in others])
             else:
                 new_dtypes = None
+            # If we have already cached the width of each column in at least
+            # one of the column's partitions, we can build new_widths for the
+            # new frame.
+            new_widths = []
+            for column in new_partitions.T:
+                found_column_width = False
+                for partition in column:
+                    if partition._width_cache is not None:
+                        new_widths.append(partition.width())
+                        found_column_width = True
+                        break
+                if not found_column_width:
+                    new_widths = None
+                    break
```

Comment:
I think this might be reduced by iterating on `part in new_partitions[0]`.

Reply:
Done.